### PR TITLE
Implements the members list export for a committee

### PR DESCRIPTION
**Pull request by romain-pierre**
Closes [EM-61](http://jira.sensiolabs.net/browse/EM-61).

### CODE REVIEW

File: src/AppBundle/Committee/CommitteeExportTransformer.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace AppBundle\Committee;
+
+use AppBundle\Entity\Adherent;
+
+class CommitteeExportTransformer
```
Comment (hhamon) [inline src/AppBundle/Committee/CommitteeExportTransformer.php]: Can't we have this class under the `Committee\Export` namespace instead?

File: src/AppBundle/Committee/CommitteeExportTransformer.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+     * @throws \BadMethodCallException
+     */
+    public static function getAdherentsAsCsv(array $adherents)
+    {
+        $handle = fopen('php://memory', 'r+');
+        fputcsv($handle, ['Prénom', 'Nom', 'Age', 'Ville', 'Pays', 'Adresse email']);
+
+        foreach ($adherents as $adherent) {
+            if (!$adherent instanceof Adherent) {
+                throw new \BadMethodCallException('This method require a collection of Adherent entities');
```
Comment (hhamon) [inline src/AppBundle/Committee/CommitteeExportTransformer.php]: `requires`.

Comment (hhamon) [inline src/AppBundle/Committee/CommitteeExportTransformer.php]: Also why not passing an ` AdherentCollection` instead to the method. This way, no need to check it contains `Adherent` instances.

File: src/AppBundle/Committee/CommitteeParametersFilter.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        return $uuids ?? [];
+    }
+
+    /**
+     * @param Uuid[]     $uuids
+     * @param Adherent[] $adherents
+     *
+     * @return Adherent[]
+     */
+    public static function removeUnknownAdherents(array $uuids, $adherents): array
```
Comment (hhamon) [inline src/AppBundle/Committee/CommitteeParametersFilter.php]: missing typehint for the second arg.

File: src/AppBundle/Committee/CommitteeExportTransformer.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+class CommitteeExportTransformer
+{
+    /**
+     * @param Adherent[] $adherents
+     *
+     * @return string
+     *
+     * @throws \BadMethodCallException
+     */
+    public static function getAdherentsAsCsv(array $adherents)
```
Comment (hhamon) [inline src/AppBundle/Committee/CommitteeExportTransformer.php]: Missing return type hint.

File: src/AppBundle/Controller/CommitteeController.php [hunk 5]
```diff
@@ -143,10 +148,37 @@ public function listMembersAction(Committee $committee): Response
     {
         $committeeManager = $this->get('app.committee_manager');
 
+        $members = $this->getDoctrine()->getRepository(CommitteeMembership::class)->findMembers($committee->getUuid());
```
Comment (hhamon) [inline src/AppBundle/Controller/CommitteeController.php]: Why not having a `findCommitteeMembers()` method to the `CommitteeManager` service class instead. It makes more sense to me.

File: src/AppBundle/Controller/CommitteeController.php [hunk 6]
```diff
@@ -143,10 +148,37 @@ public function listMembersAction(Committee $committee): Response
...
+    {
+        if (!$this->isCsrfTokenValid('committee.export_members', $request->request->get('token'))) {
+            throw $this->createAccessDeniedException('Invalid CSRF protection token to export members.');
+        }
+
+        /** @var Uuid[] $uuids */
+        $uuids = CommitteeParametersFilter::getUuidsFromJson($request->request->get('exports', ''));
+
+        /** @var Adherent[] $adherents */
+        $adherents = $this->getDoctrine()->getRepository(CommitteeMembership::class)->findMembers($committee->getUuid());
```
Comment (hhamon) [inline src/AppBundle/Controller/CommitteeController.php]: Members list must be fetched with the `CommitteeManager` service.

File: src/AppBundle/Controller/CommitteeController.php [hunk 7]
```diff
@@ -143,10 +148,37 @@ public function listMembersAction(Committee $committee): Response
...
+        $adherents = CommitteeParametersFilter::removeUnknownAdherents($uuids, $adherents);
+
+        return new Response(CommitteeExportTransformer::getAdherentsAsCsv($adherents ?? []), 200, [
+            'Content-Type' => 'application/force-download',
```
Comment (hhamon) [inline src/AppBundle/Controller/CommitteeController.php]: `Content-Type` must be `text/csv`.

File: src/AppBundle/Committee/CommitteeExportTransformer.php (hunk 2 above)
Comment (romain-pierre) [inline src/AppBundle/Committee/CommitteeExportTransformer.php]: I don't want to force to pass an instance of `AdherentCollection` but allow also a simple array or any traversable object.

File: src/AppBundle/Committee/CommitteeParametersFilter.php (hunk 3 above)
Comment (romain-pierre) [inline src/AppBundle/Committee/CommitteeParametersFilter.php]: This allow to pass array or any traversable object.

**Comment by hhamon** [comment]
Can you add unit and functional tests as well please?

**Comment by romain-pierre** [comment]
Tests are in the EM-62 issue.

**Comment by hhamon** [comment]
Ok then. Can we merge?

**Pull request merged**